### PR TITLE
nixos/loki: Implement configuration verification

### LINKED ISSUE #293088: services/loki: Configuration verification
### Describe the bug

loki's configuration should be verified by the derivation before deploying with the `-verify-config -config.file` parameters.

### Notify maintainers

@WilliButz @globin @mmahut @emilylange @ajs124 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by jpds**
## Description of changes

Fixes: #293088

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/monitoring/loki.nix [hunk 1]
```diff
@@ -99,9 +99,16 @@ in {
         conf = if cfg.configFile == null
                then prettyJSON cfg.configuration
                else cfg.configFile;
+        validateConfig = file:
+        pkgs.runCommand "validate-loki-conf" {
+          nativeBuildInputs = [ pkgs.grafana-loki ];
```
Comment (emilylange) [inline nixos/modules/services/monitoring/loki.nix:104]: This should use `cfg.package` not `pkgs.grafana-loki` :(

Comment (drupol) [reply nixos/modules/services/monitoring/loki.nix:104]: I'll provide the PR to fix this, I should have seen this.

Comment (drupol) [reply nixos/modules/services/monitoring/loki.nix:104]: #312294

**Comment by mjm** [comment]
Can we make this optional? If you use `-config.expand-env=true` to expand environment variables in your config, the validation may end up failing even though it will be fine at runtime:

```
failed parsing config: /nix/store/85dzsqn3qjhzqfjmaiz95xgr4x9jag4c-loki-config.json: parse "http://${AWS_ACCESS_KEY_ID}:${AWS_SECRET_ACCESS_KEY}@garage.service.consul:3902": net/url: invalid userinfo. Use `-config.expand-env=true` flag if you want to expand environment variables in your config file
```

**Comment by emilylange** [comment]
Thanks for reporting!

Taking a second look it's not just `services.loki.extraFlags` that breaks it.
The use of `services.loki.configFile` will error out as well.

`services.loki.configFile` is a string of a path that is not copied to the build sandbox.
It may only exist on the destination host, or only after the action script, etc.

I am tempted to add a conditional to only ever run that config validation/verification step if both `cfg.extraFlags == []`  and `cfg.configFile == null` are `true`.
I can't think of other cases were the validation step would fail.

Not sure if you are asking for a boolean option in `services.loki`.
But I think I would prefer it to have it somewhat "automagical".

Will open a PR with my proposed fix shortly.

**Comment by mjm** [comment]
I was thinking of a boolean option, similar to `services.prometheus.checkConfig`. But your automatic suggestion would work for me just as well.

**Comment by emilylange** [comment]
- #312785

**Pull request merged**